### PR TITLE
Optimization: Changing survey text without changing types shouldn't change the schema

### PR DESCRIPTION
**Pull request by DwayneJengSage**
This is for https://sagebionetworks.jira.com/browse/BRIDGE-706

This allows researchers to change survey question text or re-order survey questions without having to create a new Upload Schema or a new Synapse table.

Testing done:
- relevant unit tests
- manual tests: create new survey, update survey with same fields in different order, update survey with new fields
- TODO: Integration tests

### CODE REVIEW

File: app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java [hunk 1]
```diff
@@ -101,7 +103,24 @@ public void setStudyIdIndex(DynamoIndexHelper studyIdIndex) {
...
+            //
+            // Dump the fieldDefLists into a set, because if we have the same fields in a different order, the schemas
+            // are compatible, and we should use the old schema too.
+            Set<UploadFieldDefinition> oldFieldDefSet = ImmutableSet.copyOf(oldUploadSchema.getFieldDefinitions());
+            Set<UploadFieldDefinition> newFieldDefSet = ImmutableSet.copyOf(fieldDefList);
+            if (oldFieldDefSet.equals(newFieldDefSet)) {
+                return oldUploadSchema;
+            }
+        } else {
+            oldRev = 0;
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java]: Trivial: how about int oldRev = 0; and remove the else clause?

File: app/org/sagebionetworks/bridge/dynamodb/DynamoUploadFieldDefinition.java [hunk 2]
```diff
@@ -45,6 +46,27 @@ public boolean isRequired() {
...
+        DynamoUploadFieldDefinition that = (DynamoUploadFieldDefinition) o;
+        return Objects.equal(required, that.required) &&
+                Objects.equal(name, that.name) &&
+                Objects.equal(type, that.type);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public int hashCode() {
+        return Objects.hashCode(name, required, type);
```
Comment (alxdarksage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoUploadFieldDefinition.java:67]: I didn't know about this!

File: app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java (hunk 1 above)
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoUploadSchemaDao.java]: I think the optimizer will optimize this away either way, but I can see this from a readability perspective. Changed.

File: app/org/sagebionetworks/bridge/dynamodb/DynamoUploadFieldDefinition.java (hunk 2 above)
Comment (DwayneJengSage) [inline app/org/sagebionetworks/bridge/dynamodb/DynamoUploadFieldDefinition.java:67]: Neither did I. Apparently, IntelliJ had a setting for generating equals() and hashCode() using Guava, so I used it, and this is what happened.

**Pull request merged**